src/app/games/diam/diam.component.ts

```diff
@@ -0,0 +1,320 @@
+                    move = new DiamMoveShift(this.selected.position, 'left');
+                } else {
+                    return this.cancelMove(DiamFailure.MUST_SHIFT_LEFT_OR_RIGHT());
+                }
+            }
+            return this.chooseMove(move, this.rules.node.gameState, null, null);
+        } else {
+            return this.cancelMove(DiamFailure.MUST_SELECT_PIECE_FIRST());
+        }
+    }
+    public async onPieceInGameClick(x: number, y: number): Promise<MGPValidation> {
+        const clickValidity: MGPValidation = this.canUserPlay('#click_' + x + '_' + y);
+        if (clickValidity.isFailure()) {
+            return this.cancelMove(clickValidity.getReason());
+        }
+        if (this.selected != null) {
+            return this.onCaseClickAfterCheck(x);
+        }
+        if (this.rules.node.gameState.getPieceAtXY(x, y).owner === this.rules.node.gameState.getCurrentPlayer()) {
+            this.selected = { type: 'pieceFromBoard', position: new Coord(x, y) };
+            this.updateViewInfo();
+            return MGPValidation.SUCCESS;
+        } else {
+            return this.cancelMove(RulesFailure.MUST_CHOOSE_PLAYER_PIECE());
+        }
+    }
+    public async onRemainingPieceClick(piece: DiamPiece): Promise<MGPValidation> {
+        const clickValidity: MGPValidation = this.canUserPlay('#piece_' + piece.owner.value + '_' + (piece.otherPieceType ? 1 : 0));
+        if (clickValidity.isFailure()) {
+            return this.cancelMove(clickValidity.getReason());
+        }
+
+        if (piece.owner === this.rules.node.gameState.getCurrentPlayer()) {
+            this.selected = { type: 'pieceFromReserve', piece };
+            this.updateViewInfo();
+            return MGPValidation.SUCCESS;
+        } else {
+            return this.cancelMove(RulesFailure.MUST_CHOOSE_PLAYER_PIECE());
+        }
+    }
+    public updateBoard(): void {
+        const lastMove: DiamMove | null = this.rules.node.move;
+        if (lastMove != null) {
+            const previousState: DiamState = this.rules.node.mother.gameState;
+            if (lastMove.isDrop()) {
+                this.lastMoved = this.getLastMovedFromDrop(lastMove, previousState);
+            } else {
+                this.lastMoved = this.getLastMovedFromShift(lastMove, previousState);
+            }
+        }
+        this.updateViewInfo();
+    }
+    private getLastMovedFromDrop(drop: DiamMoveDrop, stateBefore: DiamState): LastMoved[] {
+        const startDrawPosition: Coord =
+            this.getDrawPositionRemainingPiece(drop.piece, stateBefore.getRemainingPiecesOf(drop.piece)-1);
+        const x: number = drop.getTarget();
+        const y: number = 3 - stateBefore.getStackHeight(drop.getTarget());
+        const endDrawPosition: Coord = this.getDrawPositionOnBoard(x, y);
+        const end: Coord = new Coord(x, y);
+        return [{
+            start: MGPOptional.empty(), end,
```

Comment:
this is not the way eh, all in one line or one line for all :D !

Reply:
allez en plus ça fait bien "start, end\nstart, end" :D Mais fait !